### ISSUE TITLE
In Blazor client side add support for index.cshtml instead of index.html

### ISSUE THREAD

**Issue by Andrzej-W** [aspnet/AspNetCore#10750]
Feature request described in the issue title.

I want to use index.cshtml in Blazor client side to dynamically build `<head>` section when page is requested by search bots. See #9128 and this answer https://github.com/aspnet/AspNetCore/issues/9128#issuecomment-483223453 from @SteveSandersonMS

**Comment by danroth27** [comment]
Hi @Andrzej-W 

It looks like I was *wrong* about prerendering not working with client-side Blazor in Preview5. It actually works fine: https://github.com/danroth27/ClientSideBlazorWithPrerendering. We are working on getting this documented: https://github.com/aspnet/AspNetCore.Docs/pull/12776.

Since I think this satisfies your request, I'm going to go ahead and close this issue.

**Comment by Andrzej-W** [comment]
Hello @danroth27
I have just cloned your repository, compiled, and 💥 Here is the stack trace:
```
System.InvalidOperationException: The SPA default page middleware could not return the default page '/index.html' because it was not found, and no other middleware handled the request.

   at Microsoft.AspNetCore.SpaServices.SpaDefaultPageMiddleware.<>c__DisplayClass0_0.<Attach>b__1(HttpContext context, Func`1 next)
   at Microsoft.AspNetCore.Builder.UseExtensions.<>c__DisplayClass0_1.<Use>b__1(HttpContext context)
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Builder.UseExtensions.<>c__DisplayClass0_2.<Use>b__2()
   at Microsoft.AspNetCore.Builder.UseExtensions.<>c__DisplayClass0_1.<Use>b__1(HttpContext context)
   at Microsoft.AspNetCore.Builder.Extensions.MapWhenMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```
To eliminate potential problems with HTTPS I have commented `app.UseHttpsRedirection();` but nothing changed.

I have VS 2019 16.2.0 Preview 1, Blazor Preview 5, and relatively fresh Windows 10 1809 with all updates (it is used only for Blazor)

```
PS C:\Users\aw> dotnet --info
Zestaw .NET Core SDK (odzwierciedlenie dowolnego pliku global.json):
 Version:   3.0.100-preview5-011568
 Commit:    b487ff10aa

Środowisko uruchomieniowe:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.0.100-preview5-011568\

Host (useful for support):
  Version: 3.0.0-preview5-27626-15
  Commit:  61f30f5a23

.NET Core SDKs installed:
  2.1.700-preview-009618 [C:\Program Files\dotnet\sdk]
  2.1.800-preview-009677 [C:\Program Files\dotnet\sdk]
  2.2.300-preview-010067 [C:\Program Files\dotnet\sdk]
  2.2.400-preview-010195 [C:\Program Files\dotnet\sdk]
  3.0.100-preview5-011568 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-preview5-19227-01 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-preview5-27626-15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0-preview5-27626-15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Comment by danroth27** [comment]
@Andrzej-W Sorry about that, I still had some changes on my machine that I hadn't pushed. Should be working now.

**Comment by Andrzej-W** [comment]
@danroth27 Thank you very much for an update. It is now *almost* perfect. I see only one problem. Browser sends request for `favicon.ico` file and it is not available. That is OK. Unfortunately server returns status code 500 instead of 404
```
System.InvalidOperationException: The SPA default page middleware could not return the default page '/index.html' because it was not found, and no other middleware handled the request.

   at Microsoft.AspNetCore.SpaServices.SpaDefaultPageMiddleware.<>c__DisplayClass0_0.<Attach>b__1(HttpContext context, Func`1 next)
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Builder.Extensions.MapWhenMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

**Comment by danroth27** [comment]
@Andrzej-W Thanks for letting me know! Should be fixed now.

**Comment by Andrzej-W** [comment]
@danroth27 Of course now everything works perfectly and we should always write web applications without dead links or missing favicon.ico. Unfortunately problem with wrong server status code remains. Let's add this to `index.razor`
```html
<img src="my-wrong-image-file-name.jpg"/>
```
and we will get HTTP status code 500 instead of 404. Server side Blazor correctly returns 404.

I have also noticed that `app.UseHttpsRedirection();` has no effect - all transmission is unencrypted. The problem is in `.vs\ClientSideBlazorWithPrerendering\config\applicationhost.config` - there is no binding for https protocol. This file is created automatically by Visual Studio when it does not exist. Now when almost all web applications are created with https encryption maybe VS should by default create this file with bindings for both protocols. Unfortunately I don't know where to submit this issue.

BTW, I have added binding for https and now everything is perfect.

Thank you Dan and all Blazor team members for your hard work. You make the revolution in web programming like The Beatles or Elvis Presley in music.

**Comment by Andrzej-W** [comment]
I created a new issue #11106 for wrong status code 500.

**Comment by Andrzej-W** [comment]
@danroth27 Could you update your ClientSideBlazorWithPrerendering repository for Preview6? I'm fighting with it without success. You are my last hope!

**Comment by danroth27** [comment]
Sure, see if this works for you: https://github.com/danroth27/ClientSideBlazorWithPrerendering

**Comment by Andrzej-W** [comment]
Thanks @danroth27  - works perfectly, even better than in Preview5. Missing resource, for example
```html
<img src="my-wrong-image-file-name.jpg"/>
```
returns status code 404 and there is no exception.

**Comment by AmarjeetBanwait** [comment]
@danroth27 There is some issue with this approach.
Repro: 
1. Run the samlple
2. Navigate to /fetchdata page
3. Refresh from browser refresh button

Issue:
An unhandled exception occurred while processing the request.
InvalidOperationException: Cannot provide a value for property 'Http' on type 'ClientSideBlazorWithPrerendering.Client.Pages.FetchData'. There is no registered service of type 'System.Net.Http.HttpClient'.

![Prerendering issue](https://user-images.githubusercontent.com/4598234/61167482-1437a280-a583-11e9-88d2-2b990487aded.png)

**Comment by danroth27** [comment]
@AmarjeetBanwait Yeah, I need to fix that in the client-side prerendering sample. The prerender happens *on the server*, which means you have to author your components to account for that. The HttpClient we provide as a service in client-side Blazor is specifcally setup for calling the server of origin, so it isn't intended for server use. Instead what I would recommend is abstracting away server or client specific concerns with services that have different implementations depending on where the app is running. See https://github.com/danroth27/BlazorDualMode for an example of this setup.

**Comment by AmarjeetBanwait** [comment]
> I would recommend is abstracting away server or client specific concerns with services that have different implementations depending on where the app is running

 I registered HttpClient in server project with BaseAddress and it works like a charm without implementing separate services.

```
            HttpClient httpClient = new HttpClient()
            {
                BaseAddress = new Uri("http://localhost:59250"),
            };
            services.AddSingleton(httpClient);
```
Well I know it doesn't make any sense to use HttpClient in same project but it will prevent us adding duplicate services for the same reason. and also duplicate services list will go on and on in big applications.

@danroth27 your thoughts on this workaround?

also we need to find some way to register all the client side startup services in server side startup.
Currently for prerendering we have to register same services in both startups.
for eg.
`services.AddBlazoredModal();`

it will work in prerender only if we add this in both startup files otherwise it will give same error as I described in previous comment

`There is no registered service of type `

There is also one workaround here to make ServiceCollectionExtension and register all the services in that.

```
    public static class ServiceCollectionExtensions
    {
        public static IServiceCollection RegisterServices(this IServiceCollection services)
        {
            return services.AddBlazoredModal();
            //register all your other services here
        }
    }
```

In this case we just have to add only one line in both startups

`services.RegisterServices();`

This is just a workaround 😉
